### PR TITLE
Make bftools symlink-able

### LINKED ISSUE #2540: [Linux] Make bftools symlink-able
I am managing our groups compute server, and I wanted to make the bftools available to all users. So I put the whole bftools folder in the `/opt` folder (I'm on Linux).
Then I symlinked all executables to `/usr/local/bin` such that 
     
    /usr/local/bin/bfconvert -> /opt/bftools/bfconvert

Now all users of this server can simply run bfconvert by typing `bfconvert`, however, this won't work because the `bfconvert` executable looks in the `BF_DIR` for `bf.sh` and 

    BF_DIR=`dirname "$0"`

which is `/usr/local/bin` now, and `bf.sh` is in `/opt/bftools`.

It might be nice to let the scripts resolve to the correct (symlink-resolved) directory, some suggestions are given [here](http://stackoverflow.com/questions/7665/how-to-resolve-symbolic-links-in-a-shell-script).

I could make a pull request for a Linux solution, but I don't have OS X to test it....

### PR DESCRIPTION
**Pull request by burgerga**
Resolves #2540

### CODE REVIEW

**Comment by rleigh-codelibre** [comment]
Please see https://github.com/ome/ome-files-cpp/blob/develop/cmake/TemplateShellWrapper.cmake.in#L74 to L100 which provides a portable way to determine the path in shell without using an external tool such as perl.  Such an approach could be used here and put in the generic `bf.sh` script rather than duplicating the logic in each script.  It would need a little tweaking but should otherwise do the job.


Regards,
Roger

**Comment by burgerga** [comment]
Thanks for your feedback, please also have a look at the discussion in #2540. 

The `pwd` approach in https://github.com/ome/ome-files-cpp/blob/develop/cmake/TemplateShellWrapper.cmake.in#L74 only works for directories not for files. 

Not duplicating the logic would be nice, however, it would necessitate symlinking `bf.sh` too, making it available in the `PATH`...

**Comment by sbesson** [comment]
On the OSX front, this PR was tested by modifying locally the [Bio-Formats Homebrew formula](https://github.com/ome/homebrew-alt/blob/master/Formula/bioformats52.rb) to consume the artifact produced by the daily merge build which includes this PR as follows:

```
sbesson@ls30630:homebrew-alt (master) $ git diff
diff --git a/Formula/bioformats52.rb b/Formula/bioformats52.rb
index a6a75c8..7c801ae 100644
--- a/Formula/bioformats52.rb
+++ b/Formula/bioformats52.rb
@@ -3,8 +3,9 @@ require 'formula'
 class Bioformats52 < Formula
   homepage 'http://www.openmicroscopy.org/site/products/bio-formats'
 
-  url 'http://downloads.openmicroscopy.org/bio-formats/5.2.1/artifacts/bioformats-5.2.1.zip'
-  sha256 '9a4b838952dabee0ebb5472d33cb059f5edd0b3bb77308790801875f5ffc0413'
+  url 'file:///Users/sbesson/Downloads/bioformats-5.2.1-DEV.zip'
+  version '5.2.1-DEV'
+  sha256 '6bedbe00a4431081e19f42c307b54d14664665d31f73d26dfb8d189a737b8f8a'
 
   depends_on :python => :build
   depends_on :ant => :build
@@ -18,10 +19,12 @@ class Bioformats52 < Formula
     rm Dir["tools/*.bat"]
 
     # Copy artifacts
-    bin.install "artifacts/bioformats_package.jar"
+    libexec.install "artifacts/bioformats_package.jar"
 
     # Copy command line-tools
-    bin.install Dir["tools/*"]
+    libexec.install Dir["tools/*"]
+
+    bin.install_symlink libexec/"showinf"
   end
   test do
     system "showinf", "-version"
```
After reinstalling with `brew reinstall bioformats`, the `showinf` executable is double symlinked:

```
sbesson@ls30630:homebrew-alt (master) $ ls -alh /usr/local/bin/showinf 
lrwxr-xr-x  1 sbesson  admin    44B 29 Aug 08:32 /usr/local/bin/showinf -> ../Cellar/bioformats52/5.2.1-DEV/bin/showinf
sbesson@ls30630:homebrew-alt (master) $ ls -alh /usr/local/Cellar/bioformats52/5.2.1-DEV/bin
total 8
drwxr-xr-x  3 sbesson  admin   102B 29 Aug 08:32 .
drwxr-xr-x  7 sbesson  admin   238B 29 Aug 08:32 ..
lrwxr-xr-x  1 sbesson  admin    18B 29 Aug 08:32 showinf -> ../libexec/showinf
sbesson@ls30630:homebrew-alt (master) $ ls -alh /usr/local/Cellar/bioformats52/5.2.1-DEV/libexec/
total 59088
drwxr-xr-x  19 sbesson  admin        646B 29 Aug 08:32 .
drwxr-xr-x   7 sbesson  admin        238B 29 Aug 08:32 ..
-rwxr-xr-x   1 sbesson  1133848969   2.0K 29 Aug 02:00 bf.sh
-rwxr-xr-x   1 sbesson  1133848969   374B 29 Aug 02:00 bfconvert
-rw-r--r--   1 sbesson  1133848969    29M 29 Aug 08:32 bioformats_package.jar
-rwxr-xr-x   1 sbesson  1133848969   2.9K 29 Aug 02:00 bump_maven_version.py
-rw-r--r--   1 sbesson  1133848969   1.0K 29 Aug 02:00 config.sh
-rwxr-xr-x   1 sbesson  1133848969   375B 29 Aug 02:00 domainlist
-rwxr-xr-x   1 sbesson  1133848969   379B 29 Aug 02:00 formatlist
-rwxr-xr-x   1 sbesson  1133848969   1.4K 29 Aug 02:00 ijview
-rw-r--r--   1 sbesson  1133848969   281B 29 Aug 02:00 logback.xml
-rwxr-xr-x   1 sbesson  1133848969   401B 29 Aug 02:00 mkfake
-rwxr-xr-x   1 sbesson  1133848969   428B 29 Aug 02:00 showinf
-rwxr-xr-x   1 sbesson  1133848969   8.9K 29 Aug 02:00 source-archive.py
-rwxr-xr-x   1 sbesson  1133848969   2.2K 29 Aug 02:00 test-build
-rwxr-xr-x   1 sbesson  1133848969   418B 29 Aug 02:00 tiffcomment
-rwxr-xr-x   1 sbesson  1133848969   1.2K 29 Aug 02:00 update_copyright
-rwxr-xr-x   1 sbesson  1133848969   356B 29 Aug 02:00 xmlindent
-rwxr-xr-x   1 sbesson  1133848969   352B 29 Aug 02:00 xmlvalid
sbesson@ls30630:homebrew-alt (master) $ /usr/local/bin/showinf -version
Version: 5.2.1-DEV
VCS revision: 38199fbc285d678665b1f1ecffdce2b16b96d4c8
Build date: 29 August 2016
```

The changes look good functionally and open the door to submitting the Bio-Formats formula to the upstream Homebrew repository with the JAR no longer shipped in the `bin` directory.

**Comment by rleigh-codelibre** [comment]
Tested on FreeBSD, Linux and MacOS X.  It works correctly with relative and absolute symlinks in all cases.

FreeBSD: Has `readlink -f`
MacOS X: No `readlink -f`; uses `perl` fallback.
Linux: Has `readlink -f`

On FreeBSD, the lack of a `bash` interpreter in the base install means it doesn't work until `bash` is installed.  There's really no need for bash in these scripts; we could covert them to plain `sh` scripts separately from this PR.

This all looks fine to merge.  Thanks.

**Comment by sbesson** [comment]
@burgerga: re https://github.com/openmicroscopy/bioformats/pull/2547/commits/b24aac336c6d984cce3e064f645ed53b57e8300a, can you give more context on the reason for removing `filemode`? Unzipping the [merge command-line tools artifacts](https://ci.openmicroscopy.org/view/Failing/job/BIOFORMATS-DEV-merge-build-win/lastSuccessfulBuild/artifact/artifacts/) a gives me a set of files with 644 permissions as expected from the [Ant documentation](https://ant.apache.org/manual/Types/zipfileset.html) and makes the scripts non executable when unzipped:

```
[sbesson@ome-c6100-1 ~]$ ls -alh bftools
total 29M
drwxr-xr-x  2 sbesson lsd 4.0K Aug 29 03:40 .
drwxr-xr-x 24 sbesson lsd 4.0K Aug 29 13:08 ..
-rw-r--r--  1 sbesson lsd 2.1K May 31  2015 bf.bat
-rw-r--r--  1 sbesson lsd  374 Aug 28 04:02 bfconvert
-rw-r--r--  1 sbesson lsd  298 Apr  7  2015 bfconvert.bat
-rw-r--r--  1 sbesson lsd 2.1K May 31  2015 bf.sh
-rw-r--r--  1 sbesson lsd  29M Aug 29 03:39 bioformats_package.jar
-rw-r--r--  1 sbesson lsd 1.2K Apr  7  2015 config.bat
-rw-r--r--  1 sbesson lsd 1.1K Apr  7  2015 config.sh
-rw-r--r--  1 sbesson lsd  375 Aug 28 04:02 domainlist
-rw-r--r--  1 sbesson lsd  299 Apr  7  2015 domainlist.bat
-rw-r--r--  1 sbesson lsd  379 Aug 28 04:02 formatlist
-rw-r--r--  1 sbesson lsd  303 Apr  7  2015 formatlist.bat
-rw-r--r--  1 sbesson lsd 1.5K Aug 28 04:02 ijview
-rw-r--r--  1 sbesson lsd 1.3K Apr  7  2015 ijview.bat
-rw-r--r--  1 sbesson lsd  281 Apr  7  2015 logback.xml
-rw-r--r--  1 sbesson lsd  401 Aug 28 04:02 mkfake
-rw-r--r--  1 sbesson lsd  319 Apr  7  2015 mkfake.bat
-rw-r--r--  1 sbesson lsd  428 Aug 28 04:02 showinf
-rw-r--r--  1 sbesson lsd  358 Apr  7  2015 showinf.bat
-rw-r--r--  1 sbesson lsd  418 Aug 28 04:02 tiffcomment
-rw-r--r--  1 sbesson lsd  348 Apr  7  2015 tiffcomment.bat
-rw-r--r--  1 sbesson lsd  356 Aug 28 04:02 xmlindent
-rw-r--r--  1 sbesson lsd  280 Apr  7  2015 xmlindent.bat
-rw-r--r--  1 sbesson lsd  352 Aug 28 04:02 xmlvalid
-rw-r--r--  1 sbesson lsd  276 Apr  7  2015 xmlvalid.bat
```

File: ant/toplevel.xml [hunk 1]
```diff
@@ -996,7 +996,7 @@ Type "ant -p" for a list of targets.
       <zipfileset dir="${bftools.dir}" includes="${bftools.files}"
         prefix="bftools"/>
       <zipfileset dir="${bftools.dir}" includes="${bftools.execfiles}"
-        prefix="bftools" filemode="751"/>
+        prefix="bftools"/>
```
Comment (rleigh-codelibre) [inline ant/toplevel.xml]: Note: Should be 755

**Comment by burgerga** [comment]
@sbesson Sorry, wrong assumption on my part, I assumed it would take the permissions from the source. Will make it 755

**Comment by sbesson** [comment]
Retested the output of the last commit under FreeBSD, CentOS and OSX. For all platforms, the tools worked as expected after symlinking `showinf` outside the extracted directory. Thanks @burgerga, merging.

**Comment by burgerga** [comment]
Uhmm, I got an email with this: https://ci.openmicroscopy.org/job/BIOFORMATS-DEV-breaking-test_images_good/547/changes
Somehow things related with PixelsHashes fail, but I don't see how...

**Comment by sbesson** [comment]
@burgerga: sorry about the noise. The jobs currently email all the last committers but this PR is completely unrelated to the failure above.

**Pull request merged**